### PR TITLE
feat: Add loading of environment variables from .env file

### PR DESCRIPTION
**Pull request by leoschleier**
This change adds rudimentary functionality for loading environment variables from a .env file. We add the `auto_env` function and expose it to the users of `pyper`. The function tries to load environment variables from a given `.env`. Furthermore, it

- tries to load a `.env` file from the current working directory if no path is given
- can overwrite existing environment variables (opt-in)
- returns a dictionary of environment all variables after loading

We add unit tests for this new functionality and complete the README file with a brief description.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,3 +1,32 @@
 # pyper
 
-Pyper is a Python package that offers powerful configuration management capabilities like the Viper package in Golang.
+Pyper is a Python package that offers powerful configuration management
```
Comment (geoffdavey) [inline README.md]: Not sure I would use the word "powerful" here, but can't explain exactly why. Perhaps because the term is subjective? Perhaps because it sounds like a marketing term?

Viper describes itself as "Viper is a complete configuration solution for Go applications". "complete" is also subjective, but somehow feels less like a fluffy marketing term.

File: poetry.lock [hunk 2]
```diff
@@ -1,7 +1,74 @@
 # This file is automatically @generated by Poetry 1.6.1 and should not be changed by hand.
-package = []
+
```
Comment (geoffdavey) [inline poetry.lock:2]: It would be cool if you could share your Poetry experiences next week. I guess the README should also describe how you're using Poetry.

File: pyproject.toml [hunk 3]
```diff
@@ -17,9 +20,11 @@ build-backend = "poetry.core.masonry.api"
 
 [tool.ruff]
 select = ["ALL"]
+ignore = ["ANN101", "ANN102"]  # Type annotations for 'self' and 'cls'
```
Comment (geoffdavey) [inline pyproject.toml:23]: Do these not work properly?

File: pyproject.toml [hunk 4]
```diff
@@ -17,9 +20,11 @@ build-backend = "poetry.core.masonry.api"
...
 fixable = ["ALL"]
-fix = true
+fix = false
```
Comment (geoffdavey) [inline pyproject.toml:25]: Thought we were defaults all on kinda people?!

File: src/pyper/env.py [hunk 5]
```diff
@@ -0,0 +1,78 @@
+"""Load environment variables from a `.env` file."""
+import os
+from pathlib import Path
+
+_EQUALS = "="
+_NEW_LINE = "\n"
```
Comment (geoffdavey) [inline src/pyper/env.py:6]: What about Windows line endings!? \r\n

File: src/pyper/env.py [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+    *,
+    path: Path | None = None,
+    overwrite: bool = False,
+) -> dict[str, str]:
+    """Populate environment variables with variables from a `.env` file.
+
+    Parameters
+    ----------
+    path : Path | None, optional
+        Path to the `.env` file. If not set, `pyper` will try to find
```
Comment (geoffdavey) [inline src/pyper/env.py:32]: Should be a note that the .env file should be utf8

File: tests/test_env.py [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+
+def test_auto_env_with_non_existing_path() -> None:
+    """Test the auto_env function with a non-existing path.
+
+    We expect that the function still returns a dictionary of
+    environment variables. Consequently, the dictionary must be non-
+    empty. Moreover, the dictionary must not contain the variables
+    provided in the test .env file.
+    """
+    e = env.auto_env(path=Path("non-existing-path"))
```
Comment (geoffdavey) [inline tests/test_env.py]: Maybe use the tempdir fixture to guarantee that the path doesn't exist.

File: tests/test_env.py [hunk 8]
```diff
@@ -0,0 +1,61 @@
+"""Test the env module."""
+import os
+from pathlib import Path
+
+import pytest
+from pyper import env
+
+_TEST_DOTENV_PATH = Path(__file__).parent / "data" / ".env"
```
Comment (geoffdavey) [inline tests/test_env.py:10]: You should also test a file with Windows line endings.

**Review by geoffdavey — COMMENTED** [review]
LGTM, but I'm unsure if I should be "approving".

File: pyproject.toml (hunk 3 above)
Comment (leoschleier) [reply pyproject.toml:23]: They are not auto-fixable. I do not always want to add `self: "Foo"` or `cls: type["Foo"]` (or rather `cls: Self`) to remove the warning. Therefore, I put them into the set of ignored rules.

File: pyproject.toml (hunk 4 above)
Comment (leoschleier) [reply pyproject.toml:25]: I changed my mind regarding auto-fixing on save due to the remove-unused-imports feature. I have the habit of hitting CTRL + S quite frequently. Therefore, when I import a new module and (due to my habit) save the change, it will remove my import before I use it. Instead, I often press CTRL + SHIFT + P to run Ruff in auto-fix mode. See:
image

File: src/pyper/env.py (hunk 5 above)
Comment (leoschleier) [reply src/pyper/env.py:6]: Thanks for noticing. I wasn't aware of that when I implemented this functionality. However, the built-in `open` function saved my a*: [Python Built-in Functions](https://docs.python.org/3/library/functions.html#open:~:text=newline%20determines%20how,the%20given%20string).

TL;DR: Windows' new lines `\r\n` are by default replaced with `\n`.

File: tests/test_env.py (hunk 8 above)
Comment (leoschleier) [reply tests/test_env.py:10]: Since I learned that the `open` function takes care of the newline conventions of different operating systems, I think this is not necessary anymore. See [our conversation](https://github.com/leoschleier/pyper/pull/3/files#r1335143175) above.

File: README.md (hunk 1 above)
Comment (leoschleier) [reply README.md]: I agree. I also had an uncomfortable feeling when GitHub Copilot put this description in. I will reformulate it.

File: poetry.lock (hunk 2 above)
Comment (leoschleier) [reply poetry.lock:2]: Good idea. In the future, I will have to figure out a better way of organizing the README content. For example, we probably don't want to put the "Development" section on PyPI.

File: src/pyper/env.py (hunk 6 above)
Comment (leoschleier) [reply src/pyper/env.py:32]: Good idea!

File: tests/test_env.py (hunk 7 above)
Comment (leoschleier) [reply tests/test_env.py]: That's clever! I will do that.

File: pyproject.toml (hunk 4 above)
Comment (geoffdavey) [reply pyproject.toml:25]: The golang extension has the same behaviour and initially I shared your opinion. However, after I got used to it, I loved this feature!

File: src/pyper/env.py (hunk 5 above)
Comment (geoffdavey) [reply src/pyper/env.py:6]: Maybe I knew this at some point, but it wasn't in my working set. Thanks!

**Review by geoffdavey — APPROVED** [review]
LGTM!
